### PR TITLE
#168781705 - Edit trip request

### PR DESCRIPTION
**Pull request by ngireric123**
[starts #168781705]

#### What does this PR do?

Update trip request 

#### Description of Task to be completed?

- added modify controller
- added request middleware
- added routes
#### How should this be manually tested?
- Clone the repository from [here](https://github.com/andela/caret-bn-backend.git)
- Browse to the repo directory
- navigate to Ft-update-request-168781705 branch
- Run `npm install`
- run `sequelize db:migrate`
- run `sequelize db:seed:all`
- run `npm run dev`

**- To Search Requests**
* If you are logged in 
Send a PATCH request to `http://localhost:port/api/v1/requests/:ID`

Postman body example
```
{
	"typeId": 3,
	"locationId": 1,
	
	"destinations": [
		{
			"id": 1,
			"reasons": "visit andela"
		},
		{
			"id":2,
		"reasons": "visit andela nigeria office"
		}
		]
}
```

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?
[#168781705](https://www.pivotaltracker.com/story/show/168781705)

#### Screenshots (if appropriate)
image

image

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Good Job @ngireric123 . Kindly make your error messages descriptive to tell the user what exactly they are doing it wrong. Also, make a validation for the params id because when I pass an incorrect id format the app breaks. 
Screen Shot 2019-10-29 at 13 04 04

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Great job @ngireric123 so far. It was a long road.

However, I suggest you have more descriptive error messages.
eg: You are not the owner of this request-id, 
or for destinations' id => You are not the owner of this destination id.
In short, more descriptive messages.

Kindly make sure you test your feature extensively, why has the **test coverage** dropped by **2.5%**?

Thanks

File: src/middlewares/inputValidation.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ import models from '../database/models';
 import responseError from '../utils/responseError';
 import strings from '../utils/stringsUtil';
 
+
```
Comment (nakiwuge) [inline src/middlewares/inputValidation.js]: Remove this line

File: src/middlewares/inputValidation.js [hunk 2]
```diff
@@ -159,4 +160,39 @@ export default class InputValidation {
     });
     validation(req, res, schema, next);
   }
+
+  static validateRequest(req, res, next) {
+    const data = req.body;
+    let minimumItems = 1;
+    let maximumItems = 30;
```
Comment (nakiwuge) [inline src/middlewares/inputValidation.js:166]: Add an empty line

File: src/middlewares/inputValidation.js [hunk 3]
```diff
@@ -159,4 +160,39 @@ export default class InputValidation {
...
+    if (data.typeId === 1 || data.typeId === 2) {
+      maximumItems = 1;
+    }
```
Comment (nakiwuge) [inline src/middlewares/inputValidation.js:170]: Add an empty line

File: src/middlewares/inputValidation.js [hunk 4]
```diff
@@ -159,4 +160,39 @@ export default class InputValidation {
...
+    if (data.typeId === 3) {
+      minimumItems = 2;
+    }
```
Comment (nakiwuge) [inline src/middlewares/inputValidation.js:174]: Add a new line

File: src/middlewares/request.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import string from '../utils/stringsUtil';
+import models from '../database/models';
+
+const { requests } = models;
+
+const request = {
+  async selectPending(req, res, next) {
+    const userId = req.user.payload.id;
+    const id = parseInt(req.params.id, 10);
+    const requester = await requests.findOne({
```
Comment (nakiwuge) [inline src/middlewares/request.js]: I assume you are finding a request, not a requester. Create a helper function for this query.

File: src/middlewares/request.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    if (request.userId !== req.user.payload.id) {
+      return responseError(res, 400, string.request.error.EDIT_YOUR_REQUEST);
+    }
+    req.request = request;
+    next();
+  },
+
+  async validateBody(req, res, next) {
+    const { id } = req.params;
+    const destination = req.body;
```
Comment (nakiwuge) [inline src/middlewares/request.js]: Are you using this cont anywhere? If not remove it.

File: src/middlewares/request.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    const destinationArray = destination.destinations;
+    destinationArray.forEach(async element => {
+      await models.destinations.findOne({
+        where: {
+          requestId: id, id: element.id
+        },
+      }).then(destination => {
+        if (!destination) {
+          return responseError(res, 403, `Destination of id ${element.id} does not belong to this request`);
+
```
Comment (nakiwuge) [inline src/middlewares/request.js]: Remove this line

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@ngireric123 good job on the implementation however kindly work on the feedback below.
-  Test `validateRequest.js` function to increase coverage.
- Ensure that a user can edit a return date
- Go through the comments I have left on your pr

**Review by alainmateso — APPROVED** [review]

**Review by nignanthomas — APPROVED** [review]

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**